pika/adapters/base_connection.py

```diff
@@ -347,23 +356,43 @@ def _handle_read(self):
             LOGGER.error('Read empty data, calling disconnect')
             return self._handle_disconnect()
 
+        #LOGGER.debug("read %s bytes", len(data))
         # Pass the data into our top level frame dispatching method
         self._on_data_available(data)
         return len(data)
 
     def _handle_write(self):
-        """Handle any outbound buffer writes that need to take place."""
+        """Try and write as much as we can, if we get blocked requeue 
+        what's left"""
         bytes_written = 0
-        if self.outbound_buffer:
-            frame = self.outbound_buffer.popleft()
-            try:
-                self.socket.sendall(frame)
-                bytes_written = len(frame)
-            except socket.timeout:
-                raise
-            except socket.error as error:
+        try:
+            while self.outbound_buffer:
+                frame = self.outbound_buffer.popleft()
+                bw = self.socket.send(frame)
+                frame = frame[bw:]
```

Comment:
Consider skipping object allocation if the entire frame was written. The check should be cheaper than an allocation unless you partial frame writes will be the norm.